cloud/openstack/os_api_endpoint.py

```diff
@@ -0,0 +1,591 @@
+    """
+    for current_endpoint_dict in current_endpoints_list:
+        if current_endpoint_dict.interface == interface:
+            return current_endpoint_dict
+    return None
+
+
+def detect_changes(is_v3, region, endpoints_list, current_endpoints_list):
+    """
+    Detect changes in endpoints.
+    :param state: state.
+    :param is_v3: True if we works with API V3.
+    :param region: new region.
+    :param enabled: new enabled indicator.
+    :param endpoints: new endpoint(s).
+    :param current_endpoints_list: previous setting : a dict (v2) or a dict list (V3).
+    :return: True if one or more changes are detected. In V3 case, each endpoint to be changed contains
+             an attribute 'action' with value 'update' or 'creation'.
+    """
+    changed = False
+    if is_v3:
+        for endpoint_dict in endpoints_list:
+            endp = get_endpoint_in_list(endpoint_dict.get('interface'), current_endpoints_list)
+            if endp:
+                endpoint_dict['endpoint_id'] = endp.id
+                if endp.url != endpoint_dict.get('url') or \
+                        endp.region != endpoint_dict.get('region') or \
+                        endp.enabled != endpoint_dict.get('enabled'):
+                    changed = True
+                    endpoint_dict['action'] = "update"
+                else:
+                    endpoint_dict['action'] = "none"
+            else:
+                changed = True
+                endpoint_dict['action'] = "creation"
+    else:
+        current_endpoint = current_endpoints_list[0]
+        for endpoint_dict in endpoints_list:
+            if endpoint_dict.get('interface') == 'public':
+                if not hasattr(current_endpoint, 'publicurl') or endpoint_dict.get('url') != current_endpoint.publicurl:
+                    changed = True
+            if endpoint_dict.get('interface') == 'internal':
+                if not hasattr(current_endpoint, 'internalurl') or endpoint_dict.get('url') != current_endpoint.internalurl:
+                    changed = True
+            if endpoint_dict.get('interface') == 'admin':
+                if not hasattr(current_endpoint, 'adminurl') or endpoint_dict.get('url') != current_endpoint.adminurl:
+                    changed = True
+        if region:
+            if not hasattr(current_endpoint, 'region') or current_endpoint.region != region:
+                changed = True
+        else:
+            if hasattr(current_endpoint, 'region'):
+                changed = True
+    return changed
+
+
+def validity_control(is_v3, endpoints_list):
+    """
+    Controls validity of endpoint arguments depending on API version.
+    :param is_v3: Is API V3.
+    :param interfaces_list: endpoints supplied.
+    :return: True if ok, else failed with message.
+    """
+    for ed in endpoints_list:
+        if is_v3 and (not ed.get('interface') or not ed.get('url') or not ed.get('region') or not ed.get('enabled')):
```

Comment:
Please make sure to keep the line length to >80 chars, though this is not enforced by the tests, this is the coding style we use for these modules.

This applies throughout